From bb32a6d312eceba1b1ed4a15d5e499c169d08f72 Mon Sep 17 00:00:00 2001
From: Hung-Yi <hung-yi.li@quantatw.com>
Date: Tue, 7 May 2024 20:33:14 +0800
Subject: [PATCH] modbus: extend rtu timeout

---
 subsys/modbus/modbus_serial.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/subsys/modbus/modbus_serial.c b/subsys/modbus/modbus_serial.c
index 06a3679804..50ff9a47c7 100644
--- a/subsys/modbus/modbus_serial.c
+++ b/subsys/modbus/modbus_serial.c
@@ -577,7 +577,7 @@ int modbus_serial_init(struct modbus_context *ctx,
 	}
 
 	if (param.serial.baud <= 38400) {
-		cfg->rtu_timeout = (numof_bits * if_delay_max) /
+		cfg->rtu_timeout = (numof_bits * if_delay_max) * 4 /
 				   param.serial.baud;
 	} else {
 		cfg->rtu_timeout = (numof_bits * if_delay_max) / 38400;
-- 
2.17.1

