diff --git a/drivers/flash/spi_nor_multi_dev.c b/drivers/flash/spi_nor_multi_dev.c
index 6595161c07..5dae866362 100644
--- a/drivers/flash/spi_nor_multi_dev.c
+++ b/drivers/flash/spi_nor_multi_dev.c
@@ -1713,6 +1713,14 @@ int spi_nor_re_init(const struct device *dev)
 	return spi_nor_configure(dev);
 }
 
+int spi_nor_set_freq(const struct device *dev, uint32_t freq)
+{
+	struct spi_nor_data *data = dev->data;
+	data->spi_cfg.frequency = freq;
+
+	return 0;
+}
+
 /**
  * @brief Initialize and configure the flash
  *
diff --git a/include/drivers/spi_nor.h b/include/drivers/spi_nor.h
index 42e13a9cff..aea16b6f96 100644
--- a/include/drivers/spi_nor.h
+++ b/include/drivers/spi_nor.h
@@ -141,6 +141,7 @@ struct spi_nor_op_info {
 
 int spi_nor_config_4byte_mode(const struct device *dev, bool en4b);
 int spi_nor_re_init(const struct device *dev);
+int spi_nor_set_freq(const struct device *dev, uint32_t freq);
 int spi_nor_erase_by_cmd(const struct device *dev, off_t addr,
 			 size_t size, uint8_t cmd);
 int spi_nor_get_erase_sz(const struct device *dev, uint8_t cmd);
